app/src/main/cpp/BrowserWorld.cpp

```diff
@@ -858,6 +858,14 @@ BrowserWorld::Draw() {
   m.context->Update();
   m.externalVR->PullBrowserState();
 
+  const uint32_t count = m.controllers->GetControllerCount();
+  for (uint32_t i = 0; i < count; ++i) {
+    float pulseStart = 0, pulseDuration = 0, pulseIntensity = 0;
+    uint64_t inputFrameID = 0;
+    m.externalVR->GetHapticState(i, inputFrameID, pulseStart, pulseDuration, pulseIntensity);
+    m.controllers->SetHapticsFeedback(i, inputFrameID, pulseDuration + pulseStart, pulseIntensity);
+  }
+
```

Comment:
We should probably only be processing these when in immersive mode since there shouldn't be any controllers available at other times. I would also rather it look something like this: `m.externalVR->SetHapticsState(m.controllers);`

Reply:
I was thinking if our UX are interested in using haptics feedback when doing interaction like keyboard typing, etc...If so, we can just call `ControllerContainer->SetHapticsFeedback()` and set parameters, this API will be implemented in OculusVR as well.

Reply:
We will probably use the controller API with UI but I would rather that for loop be in the externalVR code and then only have it called when in immersive mode.